test/test_mkldnn.py

```diff
@@ -122,6 +123,63 @@ def test_conv2d(self):
                 self._test_serialization(mkldnn_conv2d, (x.to_mkldnn(),))
                 self._test_tracing(mkldnn_conv2d, (x.to_mkldnn(),))
 
+    def test_conv2d_ext(self):
+        def subtensor(tensor, dim, groups, g):
+            if tensor is None:
+                return None
+            group_size = int(tensor.size(dim) / group)
+            return tensor.narrow(dim, group_size * g, group_size).contiguous()
+
+        def thnn_conv(input, weight, k, bias, s, p, d):
+            if d[0] > 1 or d[1] > 1:
+                return torch._C._nn.slow_conv_dilated2d(input, weight, k, bias, s, p, d)
+            else:
+                return torch._C._nn.thnn_conv2d(input, weight, k, bias, s, p)
+
+        def thnn_conv_group(input, weight, k, bias, s, p, d, group):
+            if group == 1:
+                return thnn_conv(input, weight, k, bias, s, p, d)
+            else:
+                outputs = []
+                for g in range(group):
+                    input_g = subtensor(input, 1, group, g)
+                    weight_g = subtensor(weight, 0, group, g)
+                    bias_g = subtensor(bias, 0, group, g)
+                    outputs.append(thnn_conv(input_g, weight_g, k, bias_g, s, p, d))
+                return torch.cat(outputs, 1)
+
+        def test_single_conv2d(g, bs, isize, ic_g, oc_g, has_bias, p, d, s, k):
+            ic, oc = ic_g * g, oc_g * g
+            input = torch.randn((bs, ic, isize[0], isize[1]), dtype=torch.float32)
+            weight = torch.randn((oc, ic_g, k[0], k[1]), dtype=torch.float32) * 0.01
+            bias = None if has_bias else torch.randn((oc), dtype=torch.float32)
+            self.assertEqual(
+                thnn_conv_group(input, weight, k, bias, s, p, d, g),
+                torch.mkldnn_convolution(input, weight, bias, p, s, d, g),
+                message='input size:{}; group:{}; input channel:{}; output channel:{};'
+                        'bias:{}, padding:{}, dilation:{}; stride:{}; kernel:{}'.format(
+                        isize, g, ic, oc, has_bias, p, d, s, k))
```

Comment:
Suggest to use relative error to check the correctness instead of `assertEqual` since we cannot guarantee `thnn_conv_group` and `mkldnn_convolution` have the same associative order.

Reply:
Oh, ignore this, `assertEqual` already checks the relative errors for tensors.